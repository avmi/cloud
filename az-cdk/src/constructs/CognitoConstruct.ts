import { Construct } from '@aws-cdk/core';
import {
  UserPool,
  SignInType,
  UserPoolAttribute,
  CfnUserPool,
  UserPoolClient,
} from '@aws-cdk/aws-cognito';

export interface ICognitoProps {
  poolName: string;
  emailSendingAccount: string;
}

export class CognitoConstruct extends Construct {
  readonly poolId: string;
  readonly clientId: string;

  constructor(scope: Construct, id: string, props: ICognitoProps) {
    super(scope, id);

    const pool = new UserPool(this, 'Pool', {
      userPoolName: props.poolName,
      signInType: SignInType.EMAIL,
      autoVerifiedAttributes: [UserPoolAttribute.EMAIL],
    });

    const client = new UserPoolClient(this, 'PoolCient', {
      userPoolClientName: `${props.poolName}-client`,
      userPool: pool,
    });

    this.poolId = pool.userPoolId;
    this.clientId = client.userPoolClientId;

    const region = pool.stack.region;
    const accountId = pool.stack.account;

    const cfn = pool.node.findChild('Resource') as CfnUserPool;

    cfn.emailConfiguration = {
      emailSendingAccount: 'DEVELOPER',
      sourceArn: `arn:aws:ses:${region}:${accountId}:identity/${props.emailSendingAccount}`,
    };
  }
}
