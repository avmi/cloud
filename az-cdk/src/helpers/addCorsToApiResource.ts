import { IResource, MockIntegration, PassthroughBehavior } from '@aws-cdk/aws-apigateway';

export const addCorsToApiResource = (
  resource: IResource,
  methods: string[] = ['GET', 'PUT', 'POST', 'DELETE'],
) => {
  const m = methods.join(',');
  resource.addMethod(
    'OPTIONS',
    new MockIntegration({
      integrationResponses: [
        {
          statusCode: '200',
          responseParameters: {
            'method.response.header.Access-Control-Allow-Headers': `'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token'`,
            'method.response.header.Access-Control-Allow-Origin': `'*'`,
            'method.response.header.Access-Control-Allow-Credentials': `'false'`,
            'method.response.header.Access-Control-Allow-Methods': `'OPTIONS,${m}'`,
          },
        },
      ],
      passthroughBehavior: PassthroughBehavior.NEVER,
      requestTemplates: {
        'application/json': '{"statusCode": 200}',
      },
    }),
    {
      methodResponses: [
        {
          statusCode: '200',
          responseParameters: {
            'method.response.header.Access-Control-Allow-Headers': true,
            'method.response.header.Access-Control-Allow-Methods': true,
            'method.response.header.Access-Control-Allow-Credentials': true,
            'method.response.header.Access-Control-Allow-Origin': true,
          },
        },
      ],
    },
  );
};
