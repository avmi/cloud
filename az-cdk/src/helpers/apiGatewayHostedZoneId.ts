/*
  The API Gateway HostedZoneId for each region has to be looked up here:
     https://docs.aws.amazon.com/general/latest/gr/rande.html#apigateway_region

  Examples:
    US East (N. Virginia)      us-east-1       execute-api.us-east-1.amazonaws.com       Z1UJRXOUMOOFQ8
    US West (N. California)    us-west-1       execute-api.us-west-1.amazonaws.com       Z2MUQ32089INYE
    US East (Ohio)             us-east-2       execute-api.us-east-2.amazonaws.com       ZOJJZC49E0EPZ
    US West (Oregon)           us-west-2       execute-api.us-west-2.amazonaws.com       Z2OJLYMUO9EFXC
    Asia Pacific (Hong Kong)   ap-east-1       execute-api.ap-east-1.amazonaws.com       Z3FD1VL90ND7K5
    Asia Pacific (Singapore)   ap-southeast-1  execute-api.ap-southeast-1.amazonaws.com  ZL327KTPIQFUL
    Asia Pacific (Sydney)      ap-southeast-2  execute-api.ap-southeast-2.amazonaws.com  Z2RPCDW04V8134
    South America (Sao Paulo)  sa-east-1       execute-api.sa-east-1.amazonaws.com       ZCMLWB8V5SYIT
*/

export const apiGatewayHostedZoneId = (region: string) => {
  switch (region.toLowerCase()) {
    case 'us-east-2':
      return 'ZOJJZC49E0EPZ';
    case 'us-east-1':
      return 'Z1UJRXOUMOOFQ8';
    case 'us-west-1':
      return 'Z2MUQ32089INYE';
    case 'us-west-2':
      return 'Z2OJLYMUO9EFXC';
    case 'ap-east-1':
      return 'Z3FD1VL90ND7K5';
    case 'ap-south-1':
      return 'Z3VO1THU9YC4UR';
    case 'ap-northeast-2':
      return 'Z20JF4UZKIW1U8';
    case 'ap-southeast-1':
      return 'ZL327KTPIQFUL';
    case 'ap-southeast-2':
      return 'Z2RPCDW04V8134';
    case 'ap-northeast-1':
      return 'Z1YSHQZHG15GKL';
    case 'ca-central-1':
      return 'Z19DQILCV0OWEC';
    case 'eu-central-1':
      return 'Z1U9ULNL0V5AJ3';
    case 'eu-west-1':
      return 'ZLY8HYME6SFDD';
    case 'eu-west-2':
      return 'ZJ5UAJN8Y3Z2Q';
    case 'eu-west-3':
      return 'Z3KY65QIEKYHQQ';
    case 'eu-north-1':
      return 'Z3UWIKFBOOGXPP';
    case 'me-south-1':
      return 'Z20ZBPC0SS8806';
    case 'sa-east-1':
      return 'ZCMLWB8V5SYIT';
  }
  throw new Error(`cannot get API Gateway HostedId for region "${region}"`);
};
