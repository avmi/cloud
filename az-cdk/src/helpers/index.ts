export * from './apiGatewayHostedZoneId';
export * from './envars2cdk';
export * from './npmCommands';
export * from './ssmSecret';
export * from './synthApp';
