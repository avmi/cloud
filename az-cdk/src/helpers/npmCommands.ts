export interface INPMCommands {
  install: string[];
  build: string[];
}

export const npmCommands = (extraBeforeTest = '', useYarn = false): INPMCommands => {
  if (useYarn) {
    return {
      install: [
        'curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -',
        'echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list',
        'sudo apt-get update && sudo apt-get install yarn',
        'yarn',
      ],
      build: ['yarn test', 'yarn build'],
    };
  }
  return {
    install: ['npm install'],
    build: extraBeforeTest
      ? [extraBeforeTest, 'npm run test', 'npm run build']
      : ['npm run test', 'npm run build'],
  };
};
