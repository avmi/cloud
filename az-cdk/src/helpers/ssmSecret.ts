import { SecretValue, CfnDynamicReference, CfnDynamicReferenceService } from '@aws-cdk/core';

// ssmSecret creates a reference to a parameter in the SSM Parameter Store (not Secrets Manager!)
export const ssmSecret = (input: { name: string; version: string }): SecretValue => {
  return SecretValue.cfnDynamicReference(
    new CfnDynamicReference(CfnDynamicReferenceService.SSM, `${input.name}:${input.version}`),
  );
};
