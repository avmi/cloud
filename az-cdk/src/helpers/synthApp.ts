import { App } from '@aws-cdk/core';

// synthApp synthetizes App and returns object
export const synthApp = (app: App): any => {
  const assembly = app.synth();
  const synthesized = assembly.stacks[0];
  return synthesized.template;
};

// synthAppString synthetizes App and returns stringifyied object
export const synthAppString = (app: App): string => {
  return JSON.stringify(synthApp(app), undefined, 2);
};
