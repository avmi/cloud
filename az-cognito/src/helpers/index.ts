export * from './flattenAttributes';
export * from './email2username';
