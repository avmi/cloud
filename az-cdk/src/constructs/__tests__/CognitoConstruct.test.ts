import { App, Stack } from '@aws-cdk/core';
import { CognitoConstruct } from '../CognitoConstruct';
import { synthAppString } from '../../helpers/synthApp';

describe('CognitoConstruct', () => {
  it('synthetizes properly', () => {
    const app = new App();
    const stack = new Stack(app, 'Stack');
    new CognitoConstruct(stack, 'Cognito', {
      poolName: 'dear-users',
      emailSendingAccount: 'tester@sender.com',
    });
    expect(synthAppString(app)).toMatchSnapshot();
  });
});

describe('CognitoConstruct + users', () => {
  it('synthetizes properly', () => {
    const app = new App();
    const stack = new Stack(app, 'Stack');
    new CognitoConstruct(stack, 'Cognito', {
      poolName: 'dear-users',
      emailSendingAccount: 'tester@sender.com',
      users: [
        {
          email: `tester@azcdk.xyz`,
          password: '<pwd-goes-here>',
          groups: 'testers',
        },
        {
          email: `tester@azcdk.xyz`,
          password: '<pwd-goes-here>',
          groups: 'visitors,travellers',
        },
      ],
    });
    expect(synthAppString(app)).toMatchSnapshot();
  });
});

describe('CognitoConstruct + Identity Providers', () => {
  it('synthetizes properly', () => {
    const app = new App();
    const stack = new Stack(app, 'Stack');
    new CognitoConstruct(stack, 'Cognito', {
      poolName: 'dear-users',
      emailSendingAccount: 'tester@sender.com',
      facebookClientId: '123412341',
      facebookClientSecret: '12341asdas12341adf',
      googleClientId: '12341324.apps.googleusercontent.com',
      googleClientSecret: '123412341',
      customDomain: 'https://myapp.custom.domain',
      customDomainCertArn: 'arn:aws:acm:us-east-1:1234',
      callbackUrls: ['https://localhost:3000/'],
      logoutUrls: ['https://localhost:3000/'],
      updateClientSettings: true,
    });
    expect(synthAppString(app)).toMatchSnapshot();
  });
});

describe('CognitoConstruct + Lambda Triggers', () => {
  it('synthetizes properly', () => {
    const app = new App();
    const stack = new Stack(app, 'Stack');
    new CognitoConstruct(stack, 'Cognito', {
      poolName: 'dear-users',
      emailSendingAccount: 'tester@sender.com',
      postConfirmTrigger: true,
      postConfirmSendEmail: true,
      postConfirmDynamoTable: 'USERS',
      dirLayers: 'src/constructs/__tests__/lambda-layers',
    });
    expect(synthAppString(app)).toMatchSnapshot();
  });
});
