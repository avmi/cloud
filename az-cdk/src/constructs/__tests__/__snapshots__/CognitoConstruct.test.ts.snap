// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CognitoConstruct + Identity Providers synthetizes properly 1`] = `
"{
  \\"Resources\\": {
    \\"CognitoPoolB9CB508B\\": {
      \\"Type\\": \\"AWS::Cognito::UserPool\\",
      \\"Properties\\": {
        \\"AutoVerifiedAttributes\\": [
          \\"email\\"
        ],
        \\"EmailConfiguration\\": {
          \\"EmailSendingAccount\\": \\"DEVELOPER\\",
          \\"SourceArn\\": {
            \\"Fn::Join\\": [
              \\"\\",
              [
                \\"arn:aws:ses:\\",
                {
                  \\"Ref\\": \\"AWS::Region\\"
                },
                \\":\\",
                {
                  \\"Ref\\": \\"AWS::AccountId\\"
                },
                \\":identity/tester@sender.com\\"
              ]
            ]
          }
        },
        \\"LambdaConfig\\": {},
        \\"UsernameAttributes\\": [
          \\"email\\"
        ],
        \\"UserPoolName\\": \\"dear-users\\"
      }
    },
    \\"CognitoPoolCient1A73565E\\": {
      \\"Type\\": \\"AWS::Cognito::UserPoolClient\\",
      \\"Properties\\": {
        \\"UserPoolId\\": {
          \\"Ref\\": \\"CognitoPoolB9CB508B\\"
        },
        \\"AllowedOAuthFlows\\": [
          \\"code\\"
        ],
        \\"AllowedOAuthFlowsUserPoolClient\\": true,
        \\"AllowedOAuthScopes\\": [
          \\"phone\\",
          \\"email\\",
          \\"openid\\",
          \\"aws.cognito.signin.user.admin\\",
          \\"profile\\"
        ],
        \\"CallbackURLs\\": [
          \\"https://localhost:3000/\\"
        ],
        \\"ClientName\\": \\"dear-users-client\\",
        \\"DefaultRedirectURI\\": \\"https://localhost:3000/\\",
        \\"GenerateSecret\\": false,
        \\"LogoutURLs\\": [
          \\"https://localhost:3000/\\"
        ],
        \\"SupportedIdentityProviders\\": [
          \\"COGNITO\\",
          \\"Facebook\\",
          \\"Google\\"
        ],
        \\"WriteAttributes\\": [
          \\"email\\",
          \\"name\\"
        ]
      }
    },
    \\"CognitoPoolDomain2AB88D72\\": {
      \\"Type\\": \\"AWS::Cognito::UserPoolDomain\\",
      \\"Properties\\": {
        \\"Domain\\": \\"https://myapp.custom.domain\\",
        \\"UserPoolId\\": {
          \\"Ref\\": \\"CognitoPoolB9CB508B\\"
        },
        \\"CustomDomainConfig\\": {
          \\"CertificateArn\\": \\"arn:aws:acm:us-east-1:1234\\"
        }
      }
    },
    \\"CognitoPoolFacebookIdP8D179BE9\\": {
      \\"Type\\": \\"AWS::Cognito::UserPoolIdentityProvider\\",
      \\"Properties\\": {
        \\"ProviderName\\": \\"Facebook\\",
        \\"ProviderType\\": \\"Facebook\\",
        \\"UserPoolId\\": {
          \\"Ref\\": \\"CognitoPoolB9CB508B\\"
        },
        \\"AttributeMapping\\": {
          \\"email\\": \\"email\\",
          \\"name\\": \\"name\\"
        },
        \\"ProviderDetails\\": {
          \\"client_id\\": \\"123412341\\",
          \\"client_secret\\": \\"12341asdas12341adf\\",
          \\"authorize_scopes\\": \\"public_profile,email\\"
        }
      }
    },
    \\"CognitoPoolGoogleIdP09966FB6\\": {
      \\"Type\\": \\"AWS::Cognito::UserPoolIdentityProvider\\",
      \\"Properties\\": {
        \\"ProviderName\\": \\"Google\\",
        \\"ProviderType\\": \\"Google\\",
        \\"UserPoolId\\": {
          \\"Ref\\": \\"CognitoPoolB9CB508B\\"
        },
        \\"AttributeMapping\\": {
          \\"email\\": \\"email\\",
          \\"name\\": \\"name\\"
        },
        \\"ProviderDetails\\": {
          \\"client_id\\": \\"12341324.apps.googleusercontent.com\\",
          \\"client_secret\\": \\"123412341\\",
          \\"authorize_scopes\\": \\"profile email openid\\"
        }
      }
    }
  }
}"
`;

exports[`CognitoConstruct + Lambda Triggers synthetizes properly 1`] = `
"{
  \\"Resources\\": {
    \\"CognitoPostConfirmServiceRole7954D5B6\\": {
      \\"Type\\": \\"AWS::IAM::Role\\",
      \\"Properties\\": {
        \\"AssumeRolePolicyDocument\\": {
          \\"Statement\\": [
            {
              \\"Action\\": \\"sts:AssumeRole\\",
              \\"Effect\\": \\"Allow\\",
              \\"Principal\\": {
                \\"Service\\": \\"lambda.amazonaws.com\\"
              }
            }
          ],
          \\"Version\\": \\"2012-10-17\\"
        },
        \\"ManagedPolicyArns\\": [
          {
            \\"Fn::Join\\": [
              \\"\\",
              [
                \\"arn:\\",
                {
                  \\"Ref\\": \\"AWS::Partition\\"
                },
                \\":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\\"
              ]
            ]
          }
        ]
      }
    },
    \\"CognitoPostConfirmServiceRoleDefaultPolicyD365441B\\": {
      \\"Type\\": \\"AWS::IAM::Policy\\",
      \\"Properties\\": {
        \\"PolicyDocument\\": {
          \\"Statement\\": [
            {
              \\"Action\\": \\"cognito-idp:*\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:SendEmail\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"dynamodb:*\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": [
                {
                  \\"Fn::Join\\": [
                    \\"\\",
                    [
                      \\"arn:aws:dynamodb:\\",
                      {
                        \\"Ref\\": \\"AWS::Region\\"
                      },
                      \\":*:table/USERS\\"
                    ]
                  ]
                },
                {
                  \\"Fn::Join\\": [
                    \\"\\",
                    [
                      \\"arn:aws:dynamodb:\\",
                      {
                        \\"Ref\\": \\"AWS::Region\\"
                      },
                      \\":*:table/USERS/index/*\\"
                    ]
                  ]
                }
              ]
            }
          ],
          \\"Version\\": \\"2012-10-17\\"
        },
        \\"PolicyName\\": \\"CognitoPostConfirmServiceRoleDefaultPolicyD365441B\\",
        \\"Roles\\": [
          {
            \\"Ref\\": \\"CognitoPostConfirmServiceRole7954D5B6\\"
          }
        ]
      }
    },
    \\"CognitoPostConfirm749659CE\\": {
      \\"Type\\": \\"AWS::Lambda::Function\\",
      \\"Properties\\": {
        \\"Code\\": {
          \\"S3Bucket\\": {
            \\"Ref\\": \\"AssetParameters705497e1a1962fcb4b819d4d8548d6ff57c279ae51687cf3db9de14432d6e30fS3Bucket95421C62\\"
          },
          \\"S3Key\\": {
            \\"Fn::Join\\": [
              \\"\\",
              [
                {
                  \\"Fn::Select\\": [
                    0,
                    {
                      \\"Fn::Split\\": [
                        \\"||\\",
                        {
                          \\"Ref\\": \\"AssetParameters705497e1a1962fcb4b819d4d8548d6ff57c279ae51687cf3db9de14432d6e30fS3VersionKey0FEDB143\\"
                        }
                      ]
                    }
                  ]
                },
                {
                  \\"Fn::Select\\": [
                    1,
                    {
                      \\"Fn::Split\\": [
                        \\"||\\",
                        {
                          \\"Ref\\": \\"AssetParameters705497e1a1962fcb4b819d4d8548d6ff57c279ae51687cf3db9de14432d6e30fS3VersionKey0FEDB143\\"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        \\"Handler\\": \\"cognitoPostConfirm.handler\\",
        \\"Role\\": {
          \\"Fn::GetAtt\\": [
            \\"CognitoPostConfirmServiceRole7954D5B6\\",
            \\"Arn\\"
          ]
        },
        \\"Runtime\\": \\"nodejs10.x\\"
      },
      \\"DependsOn\\": [
        \\"CognitoPostConfirmServiceRoleDefaultPolicyD365441B\\",
        \\"CognitoPostConfirmServiceRole7954D5B6\\"
      ]
    },
    \\"CognitoPostConfirmPostConfirmationCognito02BB4149\\": {
      \\"Type\\": \\"AWS::Lambda::Permission\\",
      \\"Properties\\": {
        \\"Action\\": \\"lambda:InvokeFunction\\",
        \\"FunctionName\\": {
          \\"Fn::GetAtt\\": [
            \\"CognitoPostConfirm749659CE\\",
            \\"Arn\\"
          ]
        },
        \\"Principal\\": \\"cognito-idp.amazonaws.com\\"
      }
    },
    \\"CognitoPoolB9CB508B\\": {
      \\"Type\\": \\"AWS::Cognito::UserPool\\",
      \\"Properties\\": {
        \\"AutoVerifiedAttributes\\": [
          \\"email\\"
        ],
        \\"EmailConfiguration\\": {
          \\"EmailSendingAccount\\": \\"DEVELOPER\\",
          \\"SourceArn\\": {
            \\"Fn::Join\\": [
              \\"\\",
              [
                \\"arn:aws:ses:\\",
                {
                  \\"Ref\\": \\"AWS::Region\\"
                },
                \\":\\",
                {
                  \\"Ref\\": \\"AWS::AccountId\\"
                },
                \\":identity/tester@sender.com\\"
              ]
            ]
          }
        },
        \\"LambdaConfig\\": {
          \\"PostConfirmation\\": {
            \\"Fn::GetAtt\\": [
              \\"CognitoPostConfirm749659CE\\",
              \\"Arn\\"
            ]
          }
        },
        \\"UsernameAttributes\\": [
          \\"email\\"
        ],
        \\"UserPoolName\\": \\"dear-users\\"
      }
    },
    \\"CognitoPoolCient1A73565E\\": {
      \\"Type\\": \\"AWS::Cognito::UserPoolClient\\",
      \\"Properties\\": {
        \\"UserPoolId\\": {
          \\"Ref\\": \\"CognitoPoolB9CB508B\\"
        },
        \\"ClientName\\": \\"dear-users-client\\",
        \\"GenerateSecret\\": false
      }
    }
  },
  \\"Parameters\\": {
    \\"AssetParameters705497e1a1962fcb4b819d4d8548d6ff57c279ae51687cf3db9de14432d6e30fS3Bucket95421C62\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"S3 bucket for asset \\\\\\"705497e1a1962fcb4b819d4d8548d6ff57c279ae51687cf3db9de14432d6e30f\\\\\\"\\"
    },
    \\"AssetParameters705497e1a1962fcb4b819d4d8548d6ff57c279ae51687cf3db9de14432d6e30fS3VersionKey0FEDB143\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"S3 key for asset version \\\\\\"705497e1a1962fcb4b819d4d8548d6ff57c279ae51687cf3db9de14432d6e30f\\\\\\"\\"
    },
    \\"AssetParameters705497e1a1962fcb4b819d4d8548d6ff57c279ae51687cf3db9de14432d6e30fArtifactHash80A7A060\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"Artifact hash for asset \\\\\\"705497e1a1962fcb4b819d4d8548d6ff57c279ae51687cf3db9de14432d6e30f\\\\\\"\\"
    }
  }
}"
`;

exports[`CognitoConstruct + users synthetizes properly 1`] = `
"{
  \\"Resources\\": {
    \\"CognitoPoolB9CB508B\\": {
      \\"Type\\": \\"AWS::Cognito::UserPool\\",
      \\"Properties\\": {
        \\"AutoVerifiedAttributes\\": [
          \\"email\\"
        ],
        \\"EmailConfiguration\\": {
          \\"EmailSendingAccount\\": \\"DEVELOPER\\",
          \\"SourceArn\\": {
            \\"Fn::Join\\": [
              \\"\\",
              [
                \\"arn:aws:ses:\\",
                {
                  \\"Ref\\": \\"AWS::Region\\"
                },
                \\":\\",
                {
                  \\"Ref\\": \\"AWS::AccountId\\"
                },
                \\":identity/tester@sender.com\\"
              ]
            ]
          }
        },
        \\"LambdaConfig\\": {},
        \\"UsernameAttributes\\": [
          \\"email\\"
        ],
        \\"UserPoolName\\": \\"dear-users\\"
      }
    },
    \\"CognitoPoolCient1A73565E\\": {
      \\"Type\\": \\"AWS::Cognito::UserPoolClient\\",
      \\"Properties\\": {
        \\"UserPoolId\\": {
          \\"Ref\\": \\"CognitoPoolB9CB508B\\"
        },
        \\"ClientName\\": \\"dear-users-client\\",
        \\"GenerateSecret\\": false
      }
    },
    \\"CognitoAddUsersFunctionServiceRoleCCEFA3E7\\": {
      \\"Type\\": \\"AWS::IAM::Role\\",
      \\"Properties\\": {
        \\"AssumeRolePolicyDocument\\": {
          \\"Statement\\": [
            {
              \\"Action\\": \\"sts:AssumeRole\\",
              \\"Effect\\": \\"Allow\\",
              \\"Principal\\": {
                \\"Service\\": \\"lambda.amazonaws.com\\"
              }
            }
          ],
          \\"Version\\": \\"2012-10-17\\"
        },
        \\"ManagedPolicyArns\\": [
          {
            \\"Fn::Join\\": [
              \\"\\",
              [
                \\"arn:\\",
                {
                  \\"Ref\\": \\"AWS::Partition\\"
                },
                \\":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\\"
              ]
            ]
          }
        ]
      }
    },
    \\"CognitoAddUsersFunctionServiceRoleDefaultPolicyAEFC94E3\\": {
      \\"Type\\": \\"AWS::IAM::Policy\\",
      \\"Properties\\": {
        \\"PolicyDocument\\": {
          \\"Statement\\": [
            {
              \\"Action\\": \\"cognito-idp:*\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            }
          ],
          \\"Version\\": \\"2012-10-17\\"
        },
        \\"PolicyName\\": \\"CognitoAddUsersFunctionServiceRoleDefaultPolicyAEFC94E3\\",
        \\"Roles\\": [
          {
            \\"Ref\\": \\"CognitoAddUsersFunctionServiceRoleCCEFA3E7\\"
          }
        ]
      }
    },
    \\"CognitoAddUsersFunctionBC0E15BB\\": {
      \\"Type\\": \\"AWS::Lambda::Function\\",
      \\"Properties\\": {
        \\"Code\\": {
          \\"S3Bucket\\": {
            \\"Ref\\": \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3Bucket84802930\\"
          },
          \\"S3Key\\": {
            \\"Fn::Join\\": [
              \\"\\",
              [
                {
                  \\"Fn::Select\\": [
                    0,
                    {
                      \\"Fn::Split\\": [
                        \\"||\\",
                        {
                          \\"Ref\\": \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3VersionKey04C3B8A9\\"
                        }
                      ]
                    }
                  ]
                },
                {
                  \\"Fn::Select\\": [
                    1,
                    {
                      \\"Fn::Split\\": [
                        \\"||\\",
                        {
                          \\"Ref\\": \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3VersionKey04C3B8A9\\"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        \\"Handler\\": \\"index.cognitoAddUsers \\",
        \\"Role\\": {
          \\"Fn::GetAtt\\": [
            \\"CognitoAddUsersFunctionServiceRoleCCEFA3E7\\",
            \\"Arn\\"
          ]
        },
        \\"Runtime\\": \\"nodejs10.x\\",
        \\"Timeout\\": 60
      },
      \\"DependsOn\\": [
        \\"CognitoAddUsersFunctionServiceRoleDefaultPolicyAEFC94E3\\",
        \\"CognitoAddUsersFunctionServiceRoleCCEFA3E7\\"
      ]
    },
    \\"CognitoAddUsers48A18D69\\": {
      \\"Type\\": \\"AWS::CloudFormation::CustomResource\\",
      \\"Properties\\": {
        \\"ServiceToken\\": {
          \\"Fn::GetAtt\\": [
            \\"CognitoAddUsersFunctionBC0E15BB\\",
            \\"Arn\\"
          ]
        },
        \\"UserPoolId\\": {
          \\"Ref\\": \\"CognitoPoolB9CB508B\\"
        },
        \\"UserPoolClientId\\": {
          \\"Ref\\": \\"CognitoPoolCient1A73565E\\"
        },
        \\"Users\\": [
          {
            \\"email\\": \\"tester@azcdk.xyz\\",
            \\"password\\": \\"<pwd-goes-here>\\",
            \\"groups\\": \\"testers\\"
          },
          {
            \\"email\\": \\"tester@azcdk.xyz\\",
            \\"password\\": \\"<pwd-goes-here>\\",
            \\"groups\\": \\"visitors,travellers\\"
          }
        ]
      },
      \\"UpdateReplacePolicy\\": \\"Delete\\",
      \\"DeletionPolicy\\": \\"Delete\\"
    }
  },
  \\"Parameters\\": {
    \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3Bucket84802930\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"S3 bucket for asset \\\\\\"8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168\\\\\\"\\"
    },
    \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3VersionKey04C3B8A9\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"S3 key for asset version \\\\\\"8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168\\\\\\"\\"
    },
    \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168ArtifactHashDB14117A\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"Artifact hash for asset \\\\\\"8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168\\\\\\"\\"
    }
  }
}"
`;

exports[`CognitoConstruct synthetizes properly 1`] = `
"{
  \\"Resources\\": {
    \\"CognitoPoolB9CB508B\\": {
      \\"Type\\": \\"AWS::Cognito::UserPool\\",
      \\"Properties\\": {
        \\"AutoVerifiedAttributes\\": [
          \\"email\\"
        ],
        \\"EmailConfiguration\\": {
          \\"EmailSendingAccount\\": \\"DEVELOPER\\",
          \\"SourceArn\\": {
            \\"Fn::Join\\": [
              \\"\\",
              [
                \\"arn:aws:ses:\\",
                {
                  \\"Ref\\": \\"AWS::Region\\"
                },
                \\":\\",
                {
                  \\"Ref\\": \\"AWS::AccountId\\"
                },
                \\":identity/tester@sender.com\\"
              ]
            ]
          }
        },
        \\"LambdaConfig\\": {},
        \\"UsernameAttributes\\": [
          \\"email\\"
        ],
        \\"UserPoolName\\": \\"dear-users\\"
      }
    },
    \\"CognitoPoolCient1A73565E\\": {
      \\"Type\\": \\"AWS::Cognito::UserPoolClient\\",
      \\"Properties\\": {
        \\"UserPoolId\\": {
          \\"Ref\\": \\"CognitoPoolB9CB508B\\"
        },
        \\"ClientName\\": \\"dear-users-client\\",
        \\"GenerateSecret\\": false
      }
    }
  }
}"
`;
